src/Microsoft.SqlTools.ServiceLayer/DacFx/DacFxService.cs

```diff
@@ -0,0 +1,257 @@
+                    ImportOperation operation = new ImportOperation(parameters, sqlConn);
+                    SqlTask sqlTask = null;
+
+                    // create task metadata
+                    TaskMetadata metadata = TaskMetadata.Create(parameters, "Import bacpac", operation, ConnectionServiceInstance);
+
+                    // put appropriate database name so that it shows imported database's name rather than master
+                    metadata.DatabaseName = parameters.TargetDatabaseName;
+
+                    sqlTask = SqlTaskManagerInstance.CreateAndRun<SqlTask>(metadata);
+
+                    await requestContext.SendResult(new ImportResult()
+                    {
+                        OperationId = operation.OperationId,
+                        Success = true,
+                        ErrorMessage = ""
+                    });
+                }
+            }
+            catch (Exception e)
+            {
+                await requestContext.SendError(e);
+            }
+        }
+
+        /// <summary>
+        /// Handles request to extract a dacpac
+        /// </summary>
+        /// <returns></returns>
+        public async Task HandleExtractRequest(ExtractParams parameters, RequestContext<ExtractResult> requestContext)
+        {
+            try
+            {
+                ConnectionInfo connInfo;
+                ConnectionServiceInstance.TryFindConnection(
+                        parameters.OwnerUri,
+                        out connInfo);
+                if (connInfo != null)
+                {
+                    SqlConnection sqlConn = ConnectionService.OpenSqlConnection(connInfo, "Extract");
+                    ExtractOperation operation = new ExtractOperation(parameters, sqlConn);
+                    SqlTask sqlTask = null;
+
+                    // create task metadata
+                    TaskMetadata metadata = TaskMetadata.Create(parameters, "Extract dacpac", operation, ConnectionServiceInstance);
+                    // put appropriate database name since connection passed was to master
+                    metadata.DatabaseName = parameters.SourceDatabaseName;
+
+                    sqlTask = SqlTaskManagerInstance.CreateAndRun<SqlTask>(metadata);
+
+                    await requestContext.SendResult(new ExtractResult()
+                    {
+                        OperationId = operation.OperationId,
+                        Success = true,
+                        ErrorMessage = ""
+                    });
+                }
+            }
+            catch (Exception e)
+            {
+                await requestContext.SendError(e);
+            }
+        }
+
+        /// <summary>
+        /// Handles request to deploy a dacpac
+        /// </summary>
+        /// <returns></returns>
+        public async Task HandleDeployRequest(DeployParams parameters, RequestContext<DeployResult> requestContext)
```

Comment:
Still seeing a lot of duplication in this HandleXXXXRequest methods. Can each handler instantiate the operation to run (ExtractOperation, DeployOperation, etc. ) and then pass the operation to a common method which handles the execution details of the operation?

Reply:
Sure I added an ExecuteOperation method that gets called by the handlers